File: iCloudDocument.m
//
//  iCloudDocument.m
//  DeepCam
//
//  Created by nichall on 08/08/2014.
//
//

#import <Foundation/Foundation.h>
#import "iCloudDocument.h"

@implementation iCloudDocument

@synthesize fileContent;

// Called whenever the application reads data from the file system
- (BOOL)loadFromContents:(id)contents ofType:(NSString *)typeName
                   error:(NSError **)outError
{
    
    if ([contents length] > 0) {
        self.fileContent = [[NSString alloc]
                            initWithBytes:[contents bytes]
                            length:[contents length]
                            encoding:NSUTF8StringEncoding];
    } else {
        // When the file is first created, assign some default content
        self.fileContent = @"Empty";
    }
    
    return YES;
}

// Called whenever the application (auto)saves the content of a file
- (id)contentsForType:(NSString *)typeName error:(NSError **)outError
{
    
    if ([self.fileContent length] == 0) {
        self.fileContent = @"Empty";
    }
    
    return [NSData dataWithBytes:[self.fileContent UTF8String]
                          length:[self.fileContent length]];
    
}

/* also see 
    http://stackoverflow.com/questions/20592884/copy-file-to-icloud-without-having-to-remove-the-local-file?rq=1

 - (void)moveFileToiCloud:(FileRepresentation *)fileToMove {
    NSURL *sourceURL = fileToMove.url;
    NSString *destinationFileName = fileToMove.fileName;
    NSURL *destinationURL = [self.documentsDir URLByAppendingPathComponent:destinationFileName];
    
    dispatch_queue_t q_default;
    q_default = dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0);
    dispatch_async(q_default, ^(void) {
        NSFileManager *fileManager = [[[NSFileManager alloc] init] autorelease];
        NSError *error = nil;
        BOOL success = [fileManager setUbiquitous:YES itemAtURL:sourceURL
                                   destinationURL:destinationURL error:&error];
        dispatch_queue_t q_main = dispatch_get_main_queue();
        dispatch_async(q_main, ^(void) {
            if (success) {
                FileRepresentation *fileRepresentation = [[FileRepresentation alloc]
                                                          initWithFileName:fileToMove.fileName url:destinationURL];
                [_fileList removeObject:fileToMove];
                [_fileList addObject:fileRepresentation];
                NSLog(@"moved file to cloud: %@", fileRepresentation);
            }
            if (!success) {
                NSLog(@"Couldn't move file to iCloud: %@", fileToMove);
            }
        });
    });
}

- (void)moveFileToLocal:(FileRepresentation *)fileToMove {
    NSURL *sourceURL = fileToMove.url;
    NSString *destinationFileName = fileToMove.fileName;
    NSURL *destinationURL = [self.documentsDir URLByAppendingPathComponent:destinationFileName];
    
    dispatch_queue_t q_default;
    q_default = dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0);
    dispatch_async(q_default, ^(void) {
        NSFileManager *fileManager = [[[NSFileManager alloc] init] autorelease];
        NSError *error = nil;
        BOOL success = [fileManager setUbiquitous:NO itemAtURL:sourceURL destinationURL:destinationURL
                                            error:&error];
        dispatch_queue_t q_main = dispatch_get_main_queue();
        dispatch_async(q_main, ^(void) {
            if (success) {
                FileRepresentation *fileRepresentation = [[FileRepresentation alloc]
                                                          initWithFileName:fileToMove.fileName url:destinationURL];
                [_fileList removeObject:fileToMove];
                [_fileList addObject:fileRepresentation];
                NSLog(@"moved file to local storage: %@", fileRepresentation);
            }
            if (!success) {
                NSLog(@"Couldn't move file to local storage: %@", fileToMove);
            }
        });
    });
}
*/
@end

File: iCloudDocument.swift
//
//  iCloudDocument.swift
//  DeepCam
//
//  Created by Nick Hall on 08/08/2014.
//  Copyright (c) 2014 cloudwise.co
//

import Foundation
import UIKit

@objc class iCloudDocument: UIDocument {
    
    var fileContent: NSString = ""
    
    init(fileURL url: NSURL!, content fileContent: NSString) {
        super.init(fileURL: url)
        self.fileContent = fileContent
    }

    // Called whenever the application reads data from the file system
    override func loadFromContents(contents: AnyObject!, ofType typeName: String!, error outError: NSErrorPointer) -> Bool
    {
        if (contents.length > 0) {
            self.fileContent = NSString(bytes: contents.bytes, length:contents.length, encoding: NSUTF8StringEncoding)
        } else {
            // When the note is first created, assign some default content
            self.fileContent = "Empty"
        }
        return true
    }
    
    override func contentsForType(typeName: String!, error outError: NSErrorPointer) -> AnyObject!
    {
    
        if (self.fileContent.length == 0) {
            self.fileContent = "Empty"
        }
        return NSData(bytes: self.fileContent.UTF8String, length: self.fileContent.length)
    }
    
    func objectiveCTestFunc () {
    
    }
}


func swiftLoadDocument()
{
    var query = NSMetadataQuery()
    
    // set the scope
    query.searchScopes = [NSMetadataQueryUbiquitousDocumentsScope]
    // define the predicate
    var pred = NSPredicate(format: "%K == %@", argumentArray: [NSMetadataItemFSNameKey, "text.txt"])
    query.predicate = pred
    // add the Finished notification
    NSNotificationCenter.defaultCenter().addObserver(nil, selector: "queryDidFinishGathering:", name:NSMetadataQueryDidFinishGatheringNotification, object:query)
    query.startQuery()
}


func swiftqueryDidFinishGathering(notification: NSNotification)
{
    let query = notification.object as NSMetadataQuery
    query.disableUpdates()
    query.stopQuery()
    
    var doc = swiftLoadData(query)
    
    NSNotificationCenter.defaultCenter().removeObserver(nil, name:NSMetadataQueryDidFinishGatheringNotification, object:query)
}


func swiftLoadData(query: NSMetadataQuery) -> UIDocument
{
    var doc: UIDocument
    
    // If the file was found, open it and return. If not found then create it and return...
    if (query.resultCount == 1)
    {
        // found the file in iCloud
        let item: AnyObject! = query.resultAtIndex(0)
        let url = item.valueForAttribute(NSMetadataItemURLKey) as NSURL
        
        doc = UIDocument(fileURL: url)
        // doc.delegate = self.viewController;
        // self.viewController.document = doc;
        
        doc.openWithCompletionHandler({success in
            if (success) {
                NSLog("loadData: existing document opened from iCloud")
            } else {
                NSLog("loadData: existing document failed to open from iCloud")
            }
            })
    } else {
        // Nothing in iCloud: create a container for file and give it URL
        NSLog("loadData: document not found in iCloud")
        
        let ubiq: NSURL = NSFileManager.defaultManager().URLForUbiquityContainerIdentifier(nil)
        let ubiquitousPackage: NSURL = ubiq.URLByAppendingPathComponent("Documents").URLByAppendingPathComponent("text.txt")
        
        doc = UIDocument(fileURL: ubiquitousPackage)
        // doc.delegate = self.viewController;
        // self.viewController.document = doc;
        
        doc.saveToURL(doc.fileURL, forSaveOperation: UIDocumentSaveOperation.ForCreating,
            completionHandler: {success in
                NSLog("loadData: new document save to iCloud")
                doc.openWithCompletionHandler({success in
                    NSLog("loadData: new document opened from iCloud")
                    })
            })
    }
    return doc
}






